csi/csi.go

```diff
@@ -257,18 +235,102 @@ func (s *OsdCsiServer) addEncryptionInfoToLabels(labels, csiSecrets map[string]s
 // Start is used to start the server.
 // It will return an error if the server is already running.
 func (s *OsdCsiServer) Start() error {
-	return s.GrpcServer.Start(func(grpcServer *grpc.Server) {
+	if err := s.GrpcServer.Start(func(grpcServer *grpc.Server) {
 		csi.RegisterIdentityServer(grpcServer, s)
 		csi.RegisterControllerServer(grpcServer, s)
 		csi.RegisterNodeServer(grpcServer, s)
-	})
+	}); err != nil {
+		return err
+	}
+
+	if s.config.Net == "unix" {
+		go func() {
+			err := autoSocketRecover(s)
```

Comment:
Looks like we'll keep checking if the socket-file exists, and keep recreating it if it doesn't.

Note, it is a "nature of Linux/Unix" to allow removing files that may be in use by other processes -- so I think there was nothing unusual with the original behavior.   Did someone requested this new _"must recreate socket-files"_ behavior?

Also, I'd be curious which `fsnotify` package did you try out?   I was aware of [dnotidy](https://en.wikipedia.org/wiki/Dnotify)/[inotify](https://en.wikipedia.org/wiki/Inotify) kernel-supported "file-watchers", but not sure if "fsnofity" uses these mechanisms or introduces something new?

Reply:
It's a common go package used in k8s a lot - https://github.com/fsnotify/fsnotify

As for the use case, there was a certain test suite that deletes the kubelet directory during a node add/removal test. Agreed this is not a normal case, but the alternative is that our CSI driver silently doesn't work until a PX restart in this case.

Reply:
Oh.. this is sweet!   Looks like that `fsnotify` is an abstraction-library over different file-system -watches, and already uses inotify for Linux.   :)

Thanks for the explanations..

Reply:
Ah cool, didn't know it uses inotify. Was wondering how it worked internally..